src/ucp/rndv/proto_rndv.c

```diff
@@ -0,0 +1,325 @@
+    ucp_md_index_t md_index;
+    ucp_proto_caps_t *caps;
+    ucp_md_map_t md_map;
+    ucs_status_t status;
+    double rts_latency;
+
+    ucs_assert(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE);
+    ucs_assert(!(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_MAX_FRAG));
+
+    /* Find lane to send the initial message */
+    rpriv->lane = ucp_proto_common_find_am_bcopy_lane(&params->super.super);
+    if (rpriv->lane == UCP_NULL_LANE) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    /* Construct select parameter for the remote protocol */
+    if (params->super.super.rkey_config_key == NULL) {
+        /* Remote buffer is unknown, assume same params as local */
+        remote_select_param          = *params->super.super.select_param;
+        remote_select_param.op_id    = params->remote_op_id;
+        remote_select_param.op_flags = 0;
+    } else {
+        /* If we know the remote buffer parameters, these are actually the local
+         * parameters for the remote protocol
+         */
+        mem_info.type    = params->super.super.rkey_config_key->mem_type;
+        mem_info.sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+        ucp_proto_select_param_init(&remote_select_param, params->remote_op_id,
+                                    0, UCP_DATATYPE_CONTIG, &mem_info, 1);
+    }
+
+    /* Initialize estimated memory registration map */
+    md_map = ucp_proto_rndv_ctrl_reg_md_map(params);
+    if (md_map == 0) {
+        ucs_trace("registration map is 0, memory type %s",
+                  ucs_memory_type_names[params->mem_info.type]);
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    rpriv->md_map           = md_map;
+    rpriv->packed_rkey_size = ucp_rkey_packed_size(context, rpriv->md_map);
+
+    /* Guess the protocol the remote side will select */
+    status = ucp_proto_rndv_ctrl_select_remote_proto(params,
+                                                     &remote_select_param,
+                                                     rpriv);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* Set send_overheads to the time to send and receive RTS message */
+    iface_attr     = ucp_proto_common_get_iface_attr(&params->super.super,
+                                                     rpriv->lane);
+    rts_latency    = (iface_attr->overhead * 2) +
+                     ucp_tl_iface_latency(context, &iface_attr->latency);
+    send_overheads = ucs_linear_func_make(rts_latency, 0.0);
+
+    /* Add registration cost to send_overheads */
+    ucs_for_each_bit(md_index, rpriv->md_map) {
+        ucs_linear_func_add_inplace(&send_overheads,
+                                    context->tl_mds[md_index].attr.reg_cost);
+    }
+
+    /* Set rendezvous protocol properties */
+    *params->super.super.priv_size         = sizeof(ucp_proto_rndv_ctrl_priv_t);
+    params->super.super.caps->cfg_thresh   = params->super.cfg_thresh;
+    params->super.super.caps->cfg_priority = params->super.cfg_priority;
```

Comment:
btw, why do we need to maintain the same parameters in both `ucp_proto_common_init_params_t` and `ucp_proto_init_params_t`?

Reply:
cfg_thresh/cfg_priority are contained in `proto_init_params_t::caps` (of type ucp_proto_caps_t) as the output of protocol initialization. In addition, they are part of `ucp_proto_common_init_params_t` as "input", for protocols which use common initialization functions. I'll try to see if we can remove them from `ucp_proto_common_init_params_t`, but maybe in subsequent PR?

Reply:
ok